### ISSUE TITLE
Unable to create Iceberg table with HiveCatalog (RuntimeMetaException: Failed to connect to Hive Metastore)

### ISSUE THREAD

**Issue by vinitamaloo-asu** [apache/iceberg#8277]
### Query engine

Spark - 3.1
Iceberg - 1.3.1

### Question

I am using HiveCatalog to create Iceberg tables with Spark as the execution engine. Everytime I try to run hiveCatalog.createTable command it gives me an exception.

This is my SparkSession configuration:

`      conf.set("spark.sql.extensions",
        "org.apache.iceberg.spark.extensions.IcebergSparkSessionExtensions")
        .set("spark.sql.catalog.spark_catalog", "org.apache.iceberg.spark.SparkSessionCatalog")
        .set("spark.sql.catalog.spark_catalog.type", "hive")
        .set("spark.sql.catalog.spark_catalog.warehouse", warehouseDir + "/spark_catalog/")
        .set("spark.sql.catalog.iceberg_catalog", "org.apache.iceberg.spark.SparkCatalog")
        .set("spark.sql.catalog.iceberg_catalog.type", "hive")
        .set("spark.sql.catalog.iceberg_catalog.warehouse", warehouseDir + "/iceberg_catalog/")
        .set("iceberg.engine.hive.enabled", "true")`

This is how I am using HiveCatalog to create table:

`    val hiveCatalog = new HiveCatalog()
    val warehouse_uri = sparkSession.sparkContext.getConf.get("spark.sql.catalog.iceberg_catalog.warehouse")
    hiveCatalog.setConf(sparkSession.sparkContext.hadoopConfiguration)

    val catalogProperties = Map(
      CatalogProperties.CATALOG_IMPL -> "org.apache.iceberg.hive.HiveCatalog",
      CatalogProperties.URI-> "",
      CatalogProperties.WAREHOUSE_LOCATION -> warehouse_uri,
      CatalogProperties.APP_ID->sparkSession.sparkContext.getConf.get("spark.app.id"),
      CatalogProperties.USER -> "###")

    hiveCatalog.initialize("iceberg_catalog", JavaConverters.mapAsJavaMap(catalogProperties))`

Below is the exception that I am seeing:

org.apache.iceberg.hive.RuntimeMetaException: Failed to connect to Hive Metastore
org.apache.iceberg.hive.HiveClientPool.newClient(HiveClientPool.java:84)
org.apache.iceberg.hive.HiveClientPool.newClient(HiveClientPool.java:34)
org.apache.iceberg.ClientPoolImpl.get(ClientPoolImpl.java:125)
org.apache.iceberg.ClientPoolImpl.run(ClientPoolImpl.java:56)
org.apache.iceberg.ClientPoolImpl.run(ClientPoolImpl.java:51)
org.apache.iceberg.hive.CachedClientPool.run(CachedClientPool.java:122)
org.apache.iceberg.hive.HiveTableOperations.doRefresh(HiveTableOperations.java:158)
org.apache.iceberg.BaseMetastoreTableOperations.refresh(BaseMetastoreTableOperations.java:97)
org.apache.iceberg.BaseMetastoreTableOperations.current(BaseMetastoreTableOperations.java:80)
org.apache.iceberg.BaseMetastoreCatalog$BaseMetastoreCatalogTableBuilder.create(BaseMetastoreCatalog.java:189)

I am using iceberg-spark-runtime-3.1_2.12 for Iceberg version 1.3.1.

Please let me know what I might be doing wrong.

**Comment by nastra** [comment]
It seems that `CatalogProperties.URI` is empty in your settings and thus it's not possible to connect to the Hive Metastore

**Comment by vinitamaloo-asu** [comment]
I am trying to connect to a local embedded metastore that's why kept this uri empty.


From my testing, I observed that when I am initializing a new HiveCatalog, it does not create HiveConf with SparkConf and only has hadoop configuration.
When SparkSession is initially configured, it creates a new HiveConf which has both sparkconf and hadoopConfiguration.

Snippet from HiveClientImpl(newHiveConf):
`    val confMap = (hadoopConf.iterator().asScala.map(kv => kv.getKey -> kv.getValue) ++
      sparkConf.getAll.toMap ++ extraConfig).toMap
    confMap.foreach { case (k, v) => hiveConf.set(k, v) }
    SQLConf.get.redactOptions(confMap).foreach { case (k, v) =>
      logDebug(s"Applying Hadoop/Hive/Spark and extra properties to Hive Conf:$k=$v")
    }`



And since my new hivecatalog does not have sparkConfig attached it also does not have all the hive configurations which I set during the sparkSession builder phase.

**Comment by vinitamaloo-asu** [comment]
Also, what is the ideal way to use HiveCatalog?
When I debugged, I observed that during sparkSession initialization, a HiveCatalog is always created for the "spark_catalog".


Is there a way to get that already initialized HiveCatalog or should I just create a HiveCatalog from scratch just like I have shared in the code?

**Comment by vinitamaloo-asu** [comment]
Does HiveCatalog even work with a local Metastore like derby? or does it just work with a remote metastore?

**Comment by RussellSpitzer** [comment]
Derby will be broken because it can't handle multithreaded access so you need either a full embedded metastore like in our tests, or a remote metastore. Or you can test with Hadoop catalog

**Comment by vinitamaloo-asu** [comment]
I observed that during sparkSession initialization, a HiveCatalog is always created for the "spark_catalog"/"iceberg_catalog"(defined by me).

I want to fetch this HiveCatalog in the app to perform DDL operations but I am not sure how to fetch it. Is there a way to 
get an instance of org.apache.iceberg.HiveCatalog for the previously initialized "iceberg_Catalog"?

Currently, I am just creating a new HiveCatalog in app with the same catalog name and properties as set in sparkSession.

**Comment by vinitamaloo-asu** [comment]
@RussellSpitzer even if I don't initialize HiveExternalCatalog, I still cannot connect to the metastore.

`
org.apache.iceberg.hive.RuntimeMetaException: Failed to connect to Hive Metastore
org.apache.iceberg.hive.HiveClientPool.newClient(HiveClientPool.java:84)
org.apache.iceberg.hive.HiveClientPool.newClient(HiveClientPool.java:34)
org.apache.iceberg.ClientPoolImpl.get(ClientPoolImpl.java:125)
org.apache.iceberg.ClientPoolImpl.run(ClientPoolImpl.java:56)
org.apache.iceberg.ClientPoolImpl.run(ClientPoolImpl.java:51)
org.apache.iceberg.hive.CachedClientPool.run(CachedClientPool.java:122)
org.apache.iceberg.hive.HiveCatalog.loadNamespaceMetadata(HiveCatalog.java:438)
org.apache.iceberg.catalog.SupportsNamespaces.namespaceExists(SupportsNamespaces.java:140)
`

Is there a way to connect the HiveCatalog to derby? If I pass uri -> "jdbc:derby:memory:metastore_749966;create=true" while initializing HiveCatalog I still get the same error.

**Comment by vinitamaloo-asu** [comment]
I was finally able to resolve this issue with the in-memory catalog implementation.
Derby didn't work with iceberg so I thought of using the file system as the database based on the suggestion [here](https://github.com/apache/iceberg/issues/7847). And to that, I had to use a Hadoop catalog(hadoop as the metastore) instead of derby.

And I also updated the test case to use HadoopCatalog instead of HiveCatalog to create tables which finally worked.
